File: src/main/java/org/simbrain/custom_sims/simulations/patterns_of_activity/TrainPredictionNet.java
package org.simbrain.custom_sims.simulations.patterns_of_activity;

import org.simbrain.network.core.NetworkUpdateAction;
import org.simbrain.network.core.Neuron;
import org.simbrain.network.core.Synapse;

/**
 * Training the prediction sub-network.
 */
public class TrainPredictionNet implements NetworkUpdateAction {

    /**
     * Reference to simulation object that has all the main variables used.
     */
    KuramotoOscillators sim;

    // TODO
    double[] lastPrediction;
    double learningRate = .1;

    /**
     * Construct the updater.
     */
    public TrainPredictionNet(KuramotoOscillators sim) {
        super();
        this.sim = sim;
        lastPrediction = sim.predictionRes.getActivations();
    }

    @Override
    public String getDescription() {
        return "Custom Learning Rule";
    }

    @Override
    public String getLongDescription() {
        return "Custom Learning Rule";
    }

    @Override
    public void invoke() {
        mainUpdateMethod();
    }

    /**
     * Training synapses using delta rule.
     */
    private void mainUpdateMethod() {

        int i = 0;
        double error = 0;
        double sumError = 0;
        for (Neuron neuron : sim.predictionRes.getNeuronList()) {
            // error = target - actual
            // error = current sensory - last prediction
            error = sim.reservoirNet.getNeuronList().get(i).getActivation() - lastPrediction[i];
            sumError += error * error;
            // System.out.println(i + ":" + error + ":" + neuron.getId());
            neuron.setAuxValue(error);
            i++;
        }

        // Update error neuron
        sim.errorNeuron.forceSetActivation(Math.sqrt(sumError));

        // Update all synapses
        for (Synapse synapse : sim.predictionSg.getAllSynapses()) {
            double newStrength = synapse.getStrength() + learningRate * synapse.getSource().getActivation() * synapse.getTarget().getAuxValue();
            synapse.setStrength(newStrength);
            // System.out.println(newStrength);
        }

        lastPrediction = sim.predictionRes.getActivations();
    }

}

File: src/main/kotlin/org/simbrain/custom_sims/simulations/patterns_of_activity/TrainPredictionNet.kt
package org.simbrain.custom_sims.simulations.patterns_of_activity

import org.simbrain.workspace.updater.UpdateAction

/**
 * Training the prediction sub-network.
 */
class TrainPredictionNet(
    /**
     * Reference to simulation object that has all the main variables used.
     */
    var sim: KuramotoOscillators
) : UpdateAction("Custom Learning Rule") {
    // TODO
    var lastPrediction: DoubleArray
    var learningRate = .1

    /**
     * Construct the updater.
     */
    init {
        lastPrediction = sim.predictionRes.activations
    }
    override suspend fun run() {
        mainUpdateMethod()
    }

    /**
     * Training synapses using delta rule.
     */
    private fun mainUpdateMethod() {
        var i = 0
        var error = 0.0
        var sumError = 0.0
        for (neuron in sim.predictionRes.neuronList) {
            // error = target - actual
            // error = current sensory - last prediction
            error = sim.reservoirNet.neuronList[i].activation - lastPrediction[i]
            sumError += error * error
            // System.out.println(i + ":" + error + ":" + neuron.getId());
            neuron.auxValue = error
            i++
        }

        // Update error neuron
        sim.errorNeuron.forceSetActivation(Math.sqrt(sumError))

        // Update all synapses
        for (synapse in sim.predictionSg.synapses) {
            val newStrength = synapse.strength + learningRate * synapse.source.activation * synapse.target.auxValue
            synapse.strength = newStrength
            // System.out.println(newStrength);
        }
        lastPrediction = sim.predictionRes.activations
    }
}